--- C:/my/mcmodding/mods/Forever-Enough-Items/overrides/ItemListOverlay.java
+++ C:/my/mcmodding/mods/Forever-Enough-Items/overrides/ItemListOverlayOverride.java
@@ -1 +1 @@
-package mezz.jei.gui;
+package code.elix_x.mods.fei.client.jeioverride;
@@ -3,3 +3,2 @@
-import javax.annotation.Nonnull;
-import javax.annotation.Nullable;
-import java.awt.*;
+import java.awt.Color;
+import java.awt.Rectangle;
@@ -7 +5,0 @@ import java.util.ArrayList;
-import java.util.Arrays;
@@ -11,0 +10,3 @@ import java.util.Objects;
+import javax.annotation.Nonnull;
+import javax.annotation.Nullable;
+
@@ -13,2 +14,4 @@ import com.google.common.collect.ImmutableList;
-import com.mojang.realmsclient.gui.ChatFormatting;
-import mezz.jei.Internal;
+
+import code.elix_x.mods.fei.ForeverEnoughItemsBase;
+import code.elix_x.mods.fei.config.FEIConfiguration;
+import code.elix_x.mods.fei.net.FEIGiveItemStackPacket;
@@ -17 +19,0 @@ import mezz.jei.JustEnoughItems;
-import mezz.jei.api.IItemListOverlay;
@@ -19 +20,0 @@ import mezz.jei.api.gui.IAdvancedGuiHandler;
-import mezz.jei.api.gui.IDrawable;
@@ -21,2 +22,4 @@ import mezz.jei.config.Config;
-import mezz.jei.config.Constants;
-import mezz.jei.config.JEIModConfigGui;
+import mezz.jei.gui.Focus;
+import mezz.jei.gui.GuiProperties;
+import mezz.jei.gui.ItemListOverlay;
+import mezz.jei.gui.TooltipRenderer;
@@ -27,4 +29,0 @@ import mezz.jei.input.GuiTextFieldFilter;
-import mezz.jei.input.ICloseable;
-import mezz.jei.input.IKeyable;
-import mezz.jei.input.IMouseHandler;
-import mezz.jei.input.IShowsRecipeFocuses;
@@ -48 +46,0 @@ import net.minecraft.util.ChatAllowedCharacters;
-import net.minecraft.util.ResourceLocation;
@@ -50 +47,0 @@ import net.minecraftforge.fml.client.config.GuiBut
-import net.minecraftforge.fml.client.config.HoverChecker;
@@ -52 +49 @@ import net.minecraftforge.fml.client.config.GuiBut
-public class ItemListOverlay implements IItemListOverlay, IShowsRecipeFocuses, IMouseHandler, IKeyable, ICloseable {
+public class ItemListOverlayOverride extends ItemListOverlay {
@@ -73,4 +69,0 @@ import net.minecraftforge.fml.client.config.GuiBut
-	private GuiButton configButton;
-	private IDrawable configButtonIcon;
-	private IDrawable configButtonCheatIcon;
-	private HoverChecker configButtonHoverChecker;
@@ -95 +88,9 @@ import net.minecraftforge.fml.client.config.GuiBut
-	public ItemListOverlay(@Nonnull ItemFilter itemFilter, @Nonnull List<IAdvancedGuiHandler<?>> advancedGuiHandlers) {
+	private boolean canGiveItems;
+	private boolean canDeleteItems;
+
+	private boolean moveSearchFieldToCenter;
+	private int searchFieldWidth;
+	private int searchFieldHeight;
+
+	public ItemListOverlayOverride(@Nonnull ItemFilter itemFilter, @Nonnull List<IAdvancedGuiHandler<?>> advancedGuiHandlers, boolean canGiveItems, boolean canDeleteItems, boolean moveSearchFieldToCenter, int searchFieldWidth, int searchFieldHeight){
+		super(itemFilter, advancedGuiHandlers);
@@ -97,0 +99,6 @@ import net.minecraftforge.fml.client.config.GuiBut
+
+		this.canGiveItems = canGiveItems;
+		this.canDeleteItems = canDeleteItems;
+		this.moveSearchFieldToCenter = moveSearchFieldToCenter;
+		this.searchFieldWidth = searchFieldWidth;
+		this.searchFieldHeight = searchFieldHeight;
@@ -100 +107,2 @@ import net.minecraftforge.fml.client.config.GuiBut
-	public void initGui(@Nonnull GuiScreen guiScreen) {
+	@Override
+	public void initGui(@Nonnull GuiScreen guiScreen){
@@ -102 +110 @@ import net.minecraftforge.fml.client.config.GuiBut
-		if (guiProperties == null) {
+		if(guiProperties == null){
@@ -108 +116 @@ import net.minecraftforge.fml.client.config.GuiBut
-		if (!activeAdvancedGuiHandlers.isEmpty() && guiScreen instanceof GuiContainer) {
+		if(!activeAdvancedGuiHandlers.isEmpty() && guiScreen instanceof GuiContainer){
@@ -111 +119 @@ import net.minecraftforge.fml.client.config.GuiBut
-		} else {
+		} else{
@@ -116 +124 @@ import net.minecraftforge.fml.client.config.GuiBut
-		if (columns < 4) {
+		if(columns < 4){
@@ -137,8 +145,13 @@ import net.minecraftforge.fml.client.config.GuiBut
-		int configButtonX = rightEdge - buttonSize + 1;
-		int configButtonY = guiProperties.getScreenHeight() - buttonSize - borderPadding;
-		configButton = new GuiButtonExt(2, configButtonX, configButtonY, buttonSize, buttonSize, null);
-		ResourceLocation configButtonIconLocation = new ResourceLocation(Constants.RESOURCE_DOMAIN, Constants.TEXTURE_GUI_PATH + "recipeBackground.png");
-		GuiHelper guiHelper = Internal.getHelpers().getGuiHelper();
-		configButtonIcon = guiHelper.createDrawable(configButtonIconLocation, 0, 166, 16, 16);
-		configButtonCheatIcon = guiHelper.createDrawable(configButtonIconLocation, 16, 166, 16, 16);
-		configButtonHoverChecker = new HoverChecker(configButton, 0);
+		if(moveSearchFieldToCenter){
+			int searchFieldY = guiProperties.getScreenHeight() - searchFieldHeight - borderPadding - 2;
+			FontRenderer fontRenderer = Minecraft.getMinecraft().fontRendererObj;
+			searchField = new GuiTextFieldFilter(0, fontRenderer, (guiProperties.getScreenWidth() - searchFieldWidth) / 2, searchFieldY, searchFieldWidth, searchFieldHeight);
+			setKeyboardFocus(false);
+			searchField.setItemFilter(itemFilter);
+		} else{
+			int searchFieldY = guiProperties.getScreenHeight() - searchFieldHeight - borderPadding - 2;
+			FontRenderer fontRenderer = Minecraft.getMinecraft().fontRendererObj;
+			searchField = new GuiTextFieldFilter(0, fontRenderer, rightEdge - searchFieldWidth - borderPadding, searchFieldY, searchFieldWidth, searchFieldHeight);
+			setKeyboardFocus(false);
+			searchField.setItemFilter(itemFilter);
+		}
@@ -146,7 +158,0 @@ import net.minecraftforge.fml.client.config.GuiBut
-		int searchFieldY = guiProperties.getScreenHeight() - searchHeight - borderPadding - 2;
-		int searchFieldWidth = rightEdge - leftEdge - buttonSize - 1;
-		FontRenderer fontRenderer = Minecraft.getMinecraft().fontRendererObj;
-		searchField = new GuiTextFieldFilter(0, fontRenderer, leftEdge, searchFieldY, searchFieldWidth, searchHeight);
-		setKeyboardFocus(false);
-		searchField.setItemFilter(itemFilter);
-
@@ -159 +165 @@ import net.minecraftforge.fml.client.config.GuiBut
-	private List<IAdvancedGuiHandler<?>> getActiveAdvancedGuiHandlers(@Nonnull GuiScreen guiScreen) {
+	private List<IAdvancedGuiHandler<?>> getActiveAdvancedGuiHandlers(@Nonnull GuiScreen guiScreen){
@@ -161 +167 @@ import net.minecraftforge.fml.client.config.GuiBut
-		if (guiScreen instanceof GuiContainer) {
+		if(guiScreen instanceof GuiContainer){
@@ -163,2 +169,2 @@ import net.minecraftforge.fml.client.config.GuiBut
-			for (IAdvancedGuiHandler<?> advancedGuiHandler : advancedGuiHandlers) {
-				if (advancedGuiHandler.getGuiContainerClass().isAssignableFrom(guiContainer.getClass())) {
+			for(IAdvancedGuiHandler<?> advancedGuiHandler : advancedGuiHandlers){
+				if(advancedGuiHandler.getGuiContainerClass().isAssignableFrom(guiContainer.getClass())){
@@ -172 +178 @@ import net.minecraftforge.fml.client.config.GuiBut
-	private List<Rectangle> getGuiAreas(GuiContainer guiContainer) {
+	private List<Rectangle> getGuiAreas(GuiContainer guiContainer){
@@ -174 +180 @@ import net.minecraftforge.fml.client.config.GuiBut
-		for (IAdvancedGuiHandler<?> advancedGuiHandler : activeAdvancedGuiHandlers) {
+		for(IAdvancedGuiHandler<?> advancedGuiHandler : activeAdvancedGuiHandlers){
@@ -176 +182 @@ import net.minecraftforge.fml.client.config.GuiBut
-			if (guiExtraAreas != null) {
+			if(guiExtraAreas != null){
@@ -183,2 +189,2 @@ import net.minecraftforge.fml.client.config.GuiBut
-	private <T extends GuiContainer> List<Rectangle> getGuiAreas(GuiContainer guiContainer, IAdvancedGuiHandler<T> advancedGuiHandler) {
-		if (advancedGuiHandler.getGuiContainerClass().isAssignableFrom(guiContainer.getClass())) {
+	private <T extends GuiContainer> List<Rectangle> getGuiAreas(GuiContainer guiContainer, IAdvancedGuiHandler<T> advancedGuiHandler){
+		if(advancedGuiHandler.getGuiContainerClass().isAssignableFrom(guiContainer.getClass())){
@@ -191,2 +197,3 @@ import net.minecraftforge.fml.client.config.GuiBut
-	public void updateGui(@Nonnull GuiScreen guiScreen) {
-		if (this.guiProperties == null) {
+	@Override
+	public void updateGui(@Nonnull GuiScreen guiScreen){
+		if(this.guiProperties == null){
@@ -194 +201 @@ import net.minecraftforge.fml.client.config.GuiBut
-		} else {
+		} else{
@@ -196 +203 @@ import net.minecraftforge.fml.client.config.GuiBut
-			if (guiProperties == null) {
+			if(guiProperties == null){
@@ -199 +206 @@ import net.minecraftforge.fml.client.config.GuiBut
-			if (!this.guiProperties.equals(guiProperties)) {
+			if(!this.guiProperties.equals(guiProperties)){
@@ -201 +208 @@ import net.minecraftforge.fml.client.config.GuiBut
-			} else if (!activeAdvancedGuiHandlers.isEmpty() && guiScreen instanceof GuiContainer) {
+			} else if(!activeAdvancedGuiHandlers.isEmpty() && guiScreen instanceof GuiContainer){
@@ -204 +211 @@ import net.minecraftforge.fml.client.config.GuiBut
-				if (!Objects.equals(this.guiAreas, guiAreas)) {
+				if(!Objects.equals(this.guiAreas, guiAreas)){
@@ -211 +218 @@ import net.minecraftforge.fml.client.config.GuiBut
-	private static void createItemButtons(@Nonnull GuiItemStackFastList guiItemStacks, @Nullable List<Rectangle> guiAreas, final int xStart, final int yStart, final int columnCount, final int rowCount) {
+	private static void createItemButtons(@Nonnull GuiItemStackFastList guiItemStacks, @Nullable List<Rectangle> guiAreas, final int xStart, final int yStart, final int columnCount, final int rowCount){
@@ -214 +221 @@ import net.minecraftforge.fml.client.config.GuiBut
-		for (int row = 0; row < rowCount; row++) {
+		for(int row = 0; row < rowCount; row++){
@@ -216 +223 @@ import net.minecraftforge.fml.client.config.GuiBut
-			for (int column = 0; column < columnCount; column++) {
+			for(int column = 0; column < columnCount; column++){
@@ -219 +226 @@ import net.minecraftforge.fml.client.config.GuiBut
-				if (guiAreas != null) {
+				if(guiAreas != null){
@@ -221 +228 @@ import net.minecraftforge.fml.client.config.GuiBut
-					if (intersects(guiAreas, stackArea)) {
+					if(intersects(guiAreas, stackArea)){
@@ -230,3 +237,3 @@ import net.minecraftforge.fml.client.config.GuiBut
-	private static boolean intersects(List<Rectangle> areas, Rectangle comparisonArea) {
-		for (Rectangle area : areas) {
-			if (area.intersects(comparisonArea)) {
+	private static boolean intersects(List<Rectangle> areas, Rectangle comparisonArea){
+		for(Rectangle area : areas){
+			if(area.intersects(comparisonArea)){
@@ -239 +246 @@ import net.minecraftforge.fml.client.config.GuiBut
-	private void updateLayout() {
+	private void updateLayout(){
@@ -253 +260 @@ import net.minecraftforge.fml.client.config.GuiBut
-	private void nextPage() {
+	private void nextPage(){
@@ -255 +262 @@ import net.minecraftforge.fml.client.config.GuiBut
-		if (itemsCount == 0) {
+		if(itemsCount == 0){
@@ -261 +268 @@ import net.minecraftforge.fml.client.config.GuiBut
-		if (firstItemIndex >= itemsCount) {
+		if(firstItemIndex >= itemsCount){
@@ -267 +274 @@ import net.minecraftforge.fml.client.config.GuiBut
-	private void previousPage() {
+	private void previousPage(){
@@ -269 +276 @@ import net.minecraftforge.fml.client.config.GuiBut
-		if (itemsPerPage == 0) {
+		if(itemsPerPage == 0){
@@ -276 +283 @@ import net.minecraftforge.fml.client.config.GuiBut
-		if (pageNum == 0) {
+		if(pageNum == 0){
@@ -278 +285 @@ import net.minecraftforge.fml.client.config.GuiBut
-		} else {
+		} else{
@@ -283 +290 @@ import net.minecraftforge.fml.client.config.GuiBut
-		if (firstItemIndex > 0 && firstItemIndex == itemsCount) {
+		if(firstItemIndex > 0 && firstItemIndex == itemsCount){
@@ -290,2 +297,3 @@ import net.minecraftforge.fml.client.config.GuiBut
-	public void drawScreen(@Nonnull Minecraft minecraft, int mouseX, int mouseY) {
-		if (!isOpen()) {
+	@Override
+	public void drawScreen(@Nonnull Minecraft minecraft, int mouseX, int mouseY){
+		if(!isOpen()){
@@ -296 +304 @@ import net.minecraftforge.fml.client.config.GuiBut
-		
+
@@ -302 +309,0 @@ import net.minecraftforge.fml.client.config.GuiBut
-		configButton.drawButton(minecraft, mouseX, mouseY);
@@ -304,3 +310,0 @@ import net.minecraftforge.fml.client.config.GuiBut
-		IDrawable icon = Config.isCheatItemsEnabled() ? configButtonCheatIcon : configButtonIcon;
-		icon.draw(minecraft, configButton.xPosition + 2, configButton.yPosition + 2);
-
@@ -309 +313 @@ import net.minecraftforge.fml.client.config.GuiBut
-		if (shouldShowDeleteItemTooltip(minecraft)) {
+		if(shouldShowDeleteItemTooltip(minecraft)){
@@ -311 +315 @@ import net.minecraftforge.fml.client.config.GuiBut
-		} else {
+		} else{
@@ -316 +320 @@ import net.minecraftforge.fml.client.config.GuiBut
-		if (hovered != null) {
+		if(hovered != null){
@@ -325,2 +329,2 @@ import net.minecraftforge.fml.client.config.GuiBut
-	private boolean shouldShowDeleteItemTooltip(Minecraft minecraft) {
-		if (Config.isDeleteItemsInCheatModeActive()) {
+	private boolean shouldShowDeleteItemTooltip(Minecraft minecraft){
+		if(canDeleteItems && FEIConfiguration.canDeleteItems(Minecraft.getMinecraft().thePlayer)){
@@ -328 +332 @@ import net.minecraftforge.fml.client.config.GuiBut
-			if (player.inventory.getItemStack() != null) {
+			if(player.inventory.getItemStack() != null){
@@ -335,2 +339,3 @@ import net.minecraftforge.fml.client.config.GuiBut
-	public void drawTooltips(@Nonnull Minecraft minecraft, int mouseX, int mouseY) {
-		if (!isOpen()) {
+	@Override
+	public void drawTooltips(@Nonnull Minecraft minecraft, int mouseX, int mouseY){
+		if(!isOpen()){
@@ -339 +344 @@ import net.minecraftforge.fml.client.config.GuiBut
-		
+
@@ -341 +346 @@ import net.minecraftforge.fml.client.config.GuiBut
-		if (mouseOver && shouldShowDeleteItemTooltip(minecraft)) {
+		if(mouseOver && shouldShowDeleteItemTooltip(minecraft)){
@@ -346 +351 @@ import net.minecraftforge.fml.client.config.GuiBut
-		if (hovered != null) {
+		if(hovered != null){
@@ -349,13 +353,0 @@ import net.minecraftforge.fml.client.config.GuiBut
-
-		if (configButtonHoverChecker.checkHover(mouseX, mouseY)) {
-			String configString = Translator.translateToLocal("jei.tooltip.config");
-			if (Config.isCheatItemsEnabled()) {
-				List<String> tooltip = Arrays.asList(
-						configString,
-						ChatFormatting.RED + Translator.translateToLocal("jei.tooltip.cheat.mode")
-				);
-				TooltipRenderer.drawHoveringText(minecraft, tooltip, mouseX, mouseY);
-			} else {
-				TooltipRenderer.drawHoveringText(minecraft, configString, mouseX, mouseY);
-			}
-		}
@@ -364,2 +356,3 @@ import net.minecraftforge.fml.client.config.GuiBut
-	public void handleTick() {
-		if (searchField != null) {
+	@Override
+	public void handleTick(){
+		if(searchField != null){
@@ -371,2 +364,2 @@ import net.minecraftforge.fml.client.config.GuiBut
-	public boolean isMouseOver(int mouseX, int mouseY) {
-		if (guiProperties == null || !isOpen() || (mouseX < guiProperties.getGuiLeft() + guiProperties.getGuiXSize())) {
+	public boolean isMouseOver(int mouseX, int mouseY){
+		if(guiProperties == null || !isOpen()){
@@ -373,0 +367,6 @@ import net.minecraftforge.fml.client.config.GuiBut
+		} else{
+			if(!(moveSearchFieldToCenter && searchField.isMouseOver(mouseX, mouseY))){
+				if(mouseX < guiProperties.getGuiLeft() + guiProperties.getGuiXSize()){
+					return false;
+				}
+			}
@@ -376,3 +375,3 @@ import net.minecraftforge.fml.client.config.GuiBut
-		if (guiAreas != null) {
-			for (Rectangle guiArea : guiAreas) {
-				if (guiArea.contains(mouseX, mouseY)) {
+		if(guiAreas != null){
+			for(Rectangle guiArea : guiAreas){
+				if(guiArea.contains(mouseX, mouseY)){
@@ -389,2 +388,2 @@ import net.minecraftforge.fml.client.config.GuiBut
-	public Focus<?> getFocusUnderMouse(int mouseX, int mouseY) {
-		if (!isMouseOver(mouseX, mouseY)) {
+	public Focus<?> getFocusUnderMouse(int mouseX, int mouseY){
+		if(!isMouseOver(mouseX, mouseY)){
@@ -395 +394,2 @@ import net.minecraftforge.fml.client.config.GuiBut
-		if (focus != null) {
+		boolean key = Thread.currentThread().getStackTrace()[2].getMethodName().equals("getFocusUnderMouseForKey");
+		if(focus != null && (key || Config.isEditModeEnabled() || !canGiveItems || !FEIConfiguration.canGiveItems(Minecraft.getMinecraft().thePlayer))){
@@ -397 +397 @@ import net.minecraftforge.fml.client.config.GuiBut
-			focus.setAllowsCheating();
+			return focus;
@@ -399 +399 @@ import net.minecraftforge.fml.client.config.GuiBut
-		return focus;
+		return null;
@@ -403 +403 @@ import net.minecraftforge.fml.client.config.GuiBut
-	public boolean canSetFocusWithMouse() {
+	public boolean canSetFocusWithMouse(){
@@ -408,2 +408,2 @@ import net.minecraftforge.fml.client.config.GuiBut
-	public boolean handleMouseClicked(int mouseX, int mouseY, int mouseButton) {
-		if (!isMouseOver(mouseX, mouseY)) {
+	public boolean handleMouseClicked(int mouseX, int mouseY, int mouseButton){
+		if(!isMouseOver(mouseX, mouseY)){
@@ -414,8 +414,7 @@ import net.minecraftforge.fml.client.config.GuiBut
-		if (Config.isDeleteItemsInCheatModeActive()) {
-			Minecraft minecraft = Minecraft.getMinecraft();
-			EntityPlayerSP player = minecraft.thePlayer;
-			ItemStack itemStack = player.inventory.getItemStack();
-			if (itemStack != null) {
-				player.inventory.setItemStack(null);
-				PacketJEI packet = new PacketDeletePlayerItem(itemStack);
-				JustEnoughItems.getProxy().sendPacketToServer(packet);
+		if(Minecraft.getMinecraft().thePlayer.inventory.getItemStack() == null){
+			Focus f = guiItemStacks.getFocusUnderMouse(mouseX, mouseY);
+			if(f != null && f.getValue() instanceof ItemStack && canGiveItems && FEIConfiguration.canGiveItems(Minecraft.getMinecraft().thePlayer)){
+				ItemStack itemstack = ((ItemStack) f.getValue()).copy();
+				if(mouseButton == 0) itemstack.stackSize = itemstack.getMaxStackSize();
+				else itemstack.stackSize = 1;
+				ForeverEnoughItemsBase.net.sendToServer(new FEIGiveItemStackPacket(itemstack));
@@ -423,0 +423,12 @@ import net.minecraftforge.fml.client.config.GuiBut
+		} else{
+			if(canDeleteItems && FEIConfiguration.canDeleteItems(Minecraft.getMinecraft().thePlayer)){
+				Minecraft minecraft = Minecraft.getMinecraft();
+				EntityPlayerSP player = minecraft.thePlayer;
+				ItemStack itemStack = player.inventory.getItemStack();
+				if(itemStack != null){
+					player.inventory.setItemStack(null);
+					PacketJEI packet = new PacketDeletePlayerItem(itemStack);
+					JustEnoughItems.getProxy().sendPacketToServer(packet);
+					return true;
+				}
+			}
@@ -427 +438 @@ import net.minecraftforge.fml.client.config.GuiBut
-		if (buttonClicked) {
+		if(buttonClicked){
@@ -436,2 +447,2 @@ import net.minecraftforge.fml.client.config.GuiBut
-	public boolean handleMouseScrolled(int mouseX, int mouseY, int scrollDelta) {
-		if (!isMouseOver(mouseX, mouseY)) {
+	public boolean handleMouseScrolled(int mouseX, int mouseY, int scrollDelta){
+		if(!isMouseOver(mouseX, mouseY)){
@@ -440 +451 @@ import net.minecraftforge.fml.client.config.GuiBut
-		if (scrollDelta < 0) {
+		if(scrollDelta < 0){
@@ -443 +454 @@ import net.minecraftforge.fml.client.config.GuiBut
-		} else if (scrollDelta > 0) {
+		} else if(scrollDelta > 0){
@@ -450 +461 @@ import net.minecraftforge.fml.client.config.GuiBut
-	private boolean handleMouseClickedButtons(int mouseX, int mouseY) {
+	private boolean handleMouseClickedButtons(int mouseX, int mouseY){
@@ -452 +463 @@ import net.minecraftforge.fml.client.config.GuiBut
-		if (nextButton.mousePressed(minecraft, mouseX, mouseY)) {
+		if(nextButton.mousePressed(minecraft, mouseX, mouseY)){
@@ -456 +467 @@ import net.minecraftforge.fml.client.config.GuiBut
-		} else if (backButton.mousePressed(minecraft, mouseX, mouseY)) {
+		} else if(backButton.mousePressed(minecraft, mouseX, mouseY)){
@@ -460,6 +470,0 @@ import net.minecraftforge.fml.client.config.GuiBut
-		} else if (configButton.mousePressed(minecraft, mouseX, mouseY)) {
-			close();
-			configButton.playPressSound(minecraft.getSoundHandler());
-			GuiScreen configScreen = new JEIModConfigGui(minecraft.currentScreen);
-			minecraft.displayGuiScreen(configScreen);
-			return true;
@@ -470 +475 @@ import net.minecraftforge.fml.client.config.GuiBut
-	private boolean handleMouseClickedSearch(int mouseX, int mouseY, int mouseButton) {
+	private boolean handleMouseClickedSearch(int mouseX, int mouseY, int mouseButton){
@@ -473 +478 @@ import net.minecraftforge.fml.client.config.GuiBut
-		if (searchClicked && searchField.handleMouseClicked(mouseX, mouseY, mouseButton)) {
+		if(searchClicked && searchField.handleMouseClicked(mouseX, mouseY, mouseButton)){
@@ -480 +485 @@ import net.minecraftforge.fml.client.config.GuiBut
-	public boolean hasKeyboardFocus() {
+	public boolean hasKeyboardFocus(){
@@ -485,2 +490,2 @@ import net.minecraftforge.fml.client.config.GuiBut
-	public void setKeyboardFocus(boolean keyboardFocus) {
-		if (searchField != null) {
+	public void setKeyboardFocus(boolean keyboardFocus){
+		if(searchField != null){
@@ -492,2 +497,2 @@ import net.minecraftforge.fml.client.config.GuiBut
-	public boolean onKeyPressed(char typedChar, int keyCode) {
-		if (hasKeyboardFocus()) {
+	public boolean onKeyPressed(char typedChar, int keyCode){
+		if(hasKeyboardFocus()){
@@ -495 +500 @@ import net.minecraftforge.fml.client.config.GuiBut
-			if (changed) {
+			if(changed){
@@ -504,2 +509,2 @@ import net.minecraftforge.fml.client.config.GuiBut
-	private int getItemButtonXSpace() {
-		if (guiProperties == null) {
+	private int getItemButtonXSpace(){
+		if(guiProperties == null){
@@ -511,2 +516,2 @@ import net.minecraftforge.fml.client.config.GuiBut
-	private int getItemButtonYSpace() {
-		if (guiProperties == null) {
+	private int getItemButtonYSpace(){
+		if(guiProperties == null){
@@ -515 +520 @@ import net.minecraftforge.fml.client.config.GuiBut
-		return guiProperties.getScreenHeight() - (buttonSize + searchHeight + 2 + (4 * borderPadding));
+		return guiProperties.getScreenHeight() - (buttonSize + (moveSearchFieldToCenter ? 0 : (searchFieldHeight + 2)) + (4 * borderPadding));
@@ -518 +523 @@ import net.minecraftforge.fml.client.config.GuiBut
-	private int getColumns() {
+	private int getColumns(){
@@ -522 +527 @@ import net.minecraftforge.fml.client.config.GuiBut
-	private int getRows() {
+	private int getRows(){
@@ -526 +531 @@ import net.minecraftforge.fml.client.config.GuiBut
-	private int getPageCount() {
+	private int getPageCount(){
@@ -529 +534 @@ import net.minecraftforge.fml.client.config.GuiBut
-		if (stacksPerPage == 0) {
+		if(stacksPerPage == 0){
@@ -537 +542 @@ import net.minecraftforge.fml.client.config.GuiBut
-	private int getPageNum() {
+	private int getPageNum(){
@@ -539 +544 @@ import net.minecraftforge.fml.client.config.GuiBut
-		if (stacksPerPage == 0) {
+		if(stacksPerPage == 0){
@@ -545 +550,2 @@ import net.minecraftforge.fml.client.config.GuiBut
-	public void open() {
+	@Override
+	public void open(){
@@ -551 +557 @@ import net.minecraftforge.fml.client.config.GuiBut
-	public void close() {
+	public void close(){
@@ -558 +564 @@ import net.minecraftforge.fml.client.config.GuiBut
-	public boolean isOpen() {
+	public boolean isOpen(){
@@ -564,2 +570,2 @@ import net.minecraftforge.fml.client.config.GuiBut
-	public ItemStack getStackUnderMouse() {
-		if (hovered == null) {
+	public ItemStack getStackUnderMouse(){
+		if(hovered == null){
@@ -567 +573 @@ import net.minecraftforge.fml.client.config.GuiBut
-		} else {
+		} else{
@@ -573,2 +579,2 @@ import net.minecraftforge.fml.client.config.GuiBut
-	public void setFilterText(@Nullable String filterText) {
-		if (filterText == null) {
+	public void setFilterText(@Nullable String filterText){
+		if(filterText == null){
@@ -581 +587 @@ import net.minecraftforge.fml.client.config.GuiBut
-		if (searchField != null) {
+		if(searchField != null){
@@ -589 +595 @@ import net.minecraftforge.fml.client.config.GuiBut
-	public String getFilterText() {
+	public String getFilterText(){
@@ -591,0 +598 @@ import net.minecraftforge.fml.client.config.GuiBut
+
